File: pilotdeck-vdi/freecad/vdi_cad_addon/engine/manifest_export.py
"""Build drawing-manifest.json from PlantModel view placements."""

from typing import Any, Dict, List, Optional


def _sheet_dims(sheet_size: str) -> Dict[str, float]:
    sizes = {
        "A1": {"width": 720.0, "height": 450.0, "symbolHalfW": 28.0, "symbolHalfH": 32.0},
        "A3": {"width": 360.0, "height": 220.0, "symbolHalfW": 24.0, "symbolHalfH": 28.0},
    }
    return sizes.get(str(sheet_size).upper(), sizes["A1"])


def build_drawing_manifest(
    plant_model: Dict[str, Any],
    drawing_number: str = "PFD-UNKNOWN",
    view_id: Optional[str] = None,
    sheet_size: str = "A1",
    symbol_scale: float = 1.0,
    fit_scale: float = 1.0,
    files: Optional[Dict[str, str]] = None,
) -> Dict[str, Any]:
    views = plant_model.get("views") or []
    view = None
    if view_id:
        view = next((v for v in views if v.get("view_id") == view_id), None)
    if not view:
        view = next((v for v in views if v.get("view_type") == "pfd"), None) or (views[0] if views else {})
    vid = view.get("view_id") or view_id or "pfd-default"

    sheet = _sheet_dims(sheet_size)
    half_w = sheet["symbolHalfW"] * symbol_scale
    half_h = sheet["symbolHalfH"] * symbol_scale
    placements = view.get("placements") or {}
    index = {o.get("object_id"): o for o in plant_model.get("objects") or [] if o.get("object_id")}

    objects: List[Dict[str, Any]] = []
    for oid, pl in placements.items():
        plant_obj = index.get(oid)
        if not plant_obj:
            continue
        x = float(pl.get("x", 0))
        y = float(pl.get("y", 0))
        w = half_w * 2
        h = half_h * 2
        bx = x - half_w
        by = y - half_h
        sw = sheet["width"]
        sh = sheet["height"]
        objects.append({
            "object_id": oid,
            "tag": plant_obj.get("tag", ""),
            "class": plant_obj.get("class", ""),
            "bbox_mm": {"x": bx, "y": by, "w": w, "h": h},
            "bbox_view": {"x": bx / sw, "y": by / sh, "w": w / sw, "h": h / sh},
        })

    return {
        "version": "1.0",
        "drawing_number": drawing_number,
        "view_id": vid,
        "sheet_size": sheet_size,
        "scale": fit_scale,
        "symbol_scale": symbol_scale,
        "sheet_mm": {"width": sheet["width"], "height": sheet["height"]},
        "objects": objects,
        "files": files or {},
    }

File: pilotdeck-vdi/mcp/vdi-cad/manifest-export.mjs
/**
 * Build drawing-manifest.json for Web/MCP pick resolution.
 */

import { getSheetFormat } from "./sheet-sizes.mjs";

/**
 * @param {object} plantModel
 * @param {object} options
 */
export function buildDrawingManifest(plantModel, options = {}) {
  const {
    drawing_number = "PFD-UNKNOWN",
    view_id = null,
    sheet_size = "A1",
    symbol_scale = 1.0,
    fit_scale = 1.0,
  } = options;

  const views = plantModel.views || [];
  const view =
    views.find((v) => v.view_id === view_id) ||
    views.find((v) => v.view_type === "pfd") ||
    views[0] ||
    {};

  const vid = view.view_id || view_id || "pfd-default";
  const sheet = getSheetFormat(sheet_size);
  const halfW = sheet.symbolHalfW * symbol_scale;
  const halfH = sheet.symbolHalfH * symbol_scale;
  const placements = view.placements || {};
  const index = new Map((plantModel.objects || []).map((o) => [o.object_id, o]));

  const objects = [];
  for (const [oid, pl] of Object.entries(placements)) {
    const plantObj = index.get(oid);
    if (!plantObj) continue;
    const x = Number(pl.x) || 0;
    const y = Number(pl.y) || 0;
    const w = halfW * 2;
    const h = halfH * 2;
    const bx = x - halfW;
    const by = y - halfH;
    objects.push({
      object_id: oid,
      tag: plantObj.tag,
      class: plantObj.class,
      bbox_mm: { x: bx, y: by, w, h },
      bbox_view: {
        x: bx / sheet.width,
        y: by / sheet.height,
        w: w / sheet.width,
        h: h / sheet.height,
      },
    });
  }

  return {
    version: "1.0",
    drawing_number,
    view_id: vid,
    sheet_size,
    scale: fit_scale,
    symbol_scale,
    sheet_mm: { width: sheet.width, height: sheet.height },
    objects,
    files: options.files || {},
  };
}

/**
 * Hit test normalized view coordinates (0–1) against manifest objects.
 */
export function resolvePickFromManifest(manifest, x, y) {
  const nx = Number(x);
  const ny = Number(y);
  if (!Number.isFinite(nx) || !Number.isFinite(ny)) {
    return { status: "error", message: "x and y must be numbers in 0–1 range" };
  }

  for (const obj of manifest.objects || []) {
    const b = obj.bbox_view;
    if (!b) continue;
    if (nx >= b.x && nx <= b.x + b.w && ny >= b.y && ny <= b.y + b.h) {
      return { status: "hit", object_id: obj.object_id, tag: obj.tag, class: obj.class, method: "bbox_view" };
    }
  }
  return { status: "miss", message: "No object at coordinates" };
}

export function manifestPathForDrawing(projectDir, drawingNumber) {
  const safe = String(drawingNumber).replace(/[^a-zA-Z0-9_-]/g, "_");
  return `${projectDir}/cad-output/${safe}-manifest.json`;
}
